### DISCUSSION TITLE
Some CAnn3D models I build!

### DISCUSSION THREAD

**Discussion by sgenoud** [donalffons/opencascade.js#84] [Show and tell]
Hei,

I have built a small layer on top of opencascade.js (that feels a bit like cadquery). It is not public yet, but it allowed me to build some nice models (that I called CAnn3D, because they are canned 3D CAD models).

First, [a box for decks of cards](https://deckinabox.sgenoud.com/quick-draw) (which is part of a set of tools to create boxes for boardgames).

I also documented some models on my blog [a gym arch](https://carrots.sgenoud.com/3d-print-a-baby-gym-arch/) and [corners to stack boxes](https://carrots.sgenoud.com/stack-homemade-cardboard-boxes/)

And also a [customisable vase](https://cadeau.sgenoud.com/wavyVase) [and another one](https://cadeau.sgenoud.com/simpleVase). As well a a [generic hanger hook](https://cadeau.sgenoud.com/hangerHook)

Would there be an interest in my (higher level that raw occt) library as well?

Generally thanks @donalffons for this great way to use opencascade!

**Comment by Irev-Dev** [comment]
Nice,

Have you seen CascadeStudio @sgenoud? Is you lib similar?

**Reply by sgenoud** [reply]
Yes, actually I found out about opencascade via cascade studio. I really liked it but there was a couple of things that did not fit with what I wanted to do:

* it is a complete environment and not just a library. CascadeStudio standard library is nice, but not easy to integrate in another project. For instance, putting it in a tool like deckinabox was not obvious. Or even including it as an iframe in blog post.
* I like to write my code in Neovim (yeah, I know this is not that important)
* I needed another way to select edges (or faces) for fillet for instance. I found the index very frustrating to use (especially because it is not topological and can change while iterating in a design

That said, a variant of CascadeStudio with my library (instead of its standard library) would be totally possible!

**Reply by sgenoud** [reply]
As an exemple of what the code looks like, this is the simple vase code:

```.js
const buildVase = (  {
    height,
    baseWidth,
    wallThickness,
    lowerCirclPosition,
    lowerCircleRadius,
    higherCircleRadius,
    higherCirclePosition,
    topRadius,
    topFillet,
    bottomHeavy,
  }
) => {

  const splinesConfig = [
    { position: lowerCirclPosition, radius: lowerCircleRadius },
    {
      position: higherCirclePosition,
      radius: higherCircleRadius,
      startFactor: bottomHeavy ? 3 : 1,
    },
    { position: 1, radius: topRadius, startFactor: bottomHeavy ? 3 : 1 },
  ];

  const sketchVaseProfile = new Sketcher("XZ")
    .movePointer(0, 0)
    .hLine(baseWidth);

  splinesConfig.forEach(({ position, radius, startFactor, endFactor }) => {
    sketchVaseProfile.smoothSplineTo([baseWidth * radius, height * position], {
      deviation: [0, 1],
      startFactor,
      endFactor,
    });
  });

  let vase = sketchVaseProfile.lineTo(0, height).close({
    returnType: "revolutionSolid",
  });

  if (wallThickness) {
    vase = vase.shell({
      filter: new FaceFinder().containsPoint([0, 0, height]),
      thickness: wallThickness,
    });
  }

  if (topFillet) {
    vase = vase.fillet({
      filter: new EdgeFinder().inPlane("XY", height),
      radius: wallThickness / 3,
    });
  }

  return vase;
}
```

**Reply by Irev-Dev** [reply]
Very interesting,

I really do like the the idea of a higher level lib built on top of opencascade.js, if for nothing more than it being brep but also to escape the CSG mindset of other code-CAD libs. I agree with your comments about CascadeStudio, I really like the project, but definitely a few things I would have done differently (running outside of the browser is important imo)

**Comment by donalffons** [comment]
That's beautiful! :slightly_smiling_face: Great showcase and the applications feel very well made. And the fact that you're using React-Three-Fiber and Leva make me feel at home :wink:.

Some ideas regarding libraries in general:

- I would love to see libraries around opencascade.js. As we all know, OpenCascade tends to be very verbose and can be quite complex. In the same time, there seems to be a lack of resources that can teach users how to use the library effectively.
- I don't think that a library should aim at completely abstracting away opencascade.js. Ideally I would like to be able to compose different libraries inside one opencascade.js application. That also means that the initialization of opencascade.js should not be part of the library.
- I think libraries should try to focus on certain aspects. Having a "scattered" library to do various unrelated things might not be a good way to go.
- Libraries should be published as NPM-packages, so that they can be easily integrated into user's projects.
- Libraries should follow common "best practives", i.e. be tree-shakeable, shouldn't mutate prototypes and ideally come as JavaScript with `.d.ts` type definitions.

I've lately also been thinking on the topic of IDE's... I think it would be fantastic if there was a way to combine the powers of a real browser based IDE like [Stackblitz](https://stackblitz.com/) (including the ability to install libraries via npm or unpkg) and a CAD editor like CascadeStudio. But this might not be an easy thing to pull off... In case @zalo has thought about this before, it would be great to hear your thoughts.

**Reply by sgenoud** [reply]
Where do you think the memory management should be done. In a library or in opencascade.js?

It seems that having the library do the garbage collection seems cleaner. It should do some of it anyway, and having a shared responsibility does not feel right.

**Reply by donalffons** [reply]
Hmm, good question. I think you're right - the library should be responsible for cleaning up everything that is considered "temporary". 

Of course, if the library returns (for example) a `TopoDS_Shape` to the user, the user then "owns" that object and the library can no longer be responsible for cleaning that up.

**Reply by zalo** [reply]
There's not a bad argument to be made for making CascadeStudio into an extension for VS Code.

As long as it adheres to [this format](https://code.visualstudio.com/api/extension-guides/web-extensions), then local online instances of VSCode (like https://github.dev/ ) should be able to hold the model rendering window alongside the code inside the IDE.

That way, you'd get to reap the ecosystem benefits of VS Code (native Github version control, other extensions, npm packages(?) etc.) without having to sacrifice local execution like Stackblitz might require...   It might also simplify the codebase a tiny bit... but I'd be sad about losing 1-click browser model sharing and the loss of accessibility (locking modelling behind VS Code and an extension installation).

**Comment by timbod7** [comment]
@sgenoud this looks really nice. Do you plan to share your library?

**Reply by sgenoud** [reply]
Yes, I want to - I am working on the API of the library to have something relatively stable from the get go!

**Reply by promontis** [reply]
Also interesting: https://github.com/pmndrs/drei/issues/390

**Reply by promontis** [reply]
ah cool... you're display some sort of animated blob when the model is being build

**Reply by sgenoud** [reply]
I am actually using drei in the viewer 😉. The animated blob appears when something broke (sometimes it is about input being not supported).

**Reply by promontis** [reply]
I want it!!!! 🤤

**Reply by promontis** [reply]
Can we help you with building the library?

**Reply by sgenoud** [reply]
Trying to build it myself, I am currently iterating on my own. My idea is to publish it dual licensed (in the spirit of highcharts).

Generally, I am making good progress. I can't resist to show the full open cascade bottle (with the thread).
It feels more readable. I am especially proud of the Sketcher / FaceSketcher interfaces (improved from CadQuery):

```.js
const bottle = (
  oc,
  { width: myWidth, height: myHeight, thickness: myThickness }
) => {
  setOC(oc);

  let shape = new Sketcher()
    .movePointerTo([-myWidth / 2, 0])
    .vLine(-myThickness / 4)
    .threePointsArc(myWidth, 0, myWidth / 2, -myThickness / 4)
    .vLine(myThickness / 4)
    .closeWithMirror()
    .extrude(myHeight)
    .fillet(myThickness / 12);

  const myNeckRadius = myThickness / 4;
  const myNeckHeight = myHeight / 10;
  const neck = makeCylinder(
    myNeckRadius,
    myNeckHeight,
    [0, 0, myHeight],
    [0, 0, 1]
  );

  shape = shape.fuse(neck);

  shape = shape.shell({
    filter: new FaceFinder().inPlane("XY", [0, 0, myHeight + myNeckHeight]),
    thickness: myThickness / 50,
  });

  const [r, gc] = localGC();

  const neckFace = r(
    new FaceFinder()
      .containsPoint([0, myNeckRadius, myHeight])
      .ofSurfaceType("CYLINDRE")
      .find(shape, { unique: true, clean: false })
  );
  const topThreadFace = r(
    r(makeOffset(neckFace, 0.05 * myNeckRadius)).faces[0]
  );

  const baseThreadSketch = new FaceSketcher(neckFace)
    .movePointerTo([0, 0.25])
    .halfEllipse(2, 0.5, 0.7)
    .close();
  const topThreadSketch = new FaceSketcher(topThreadFace)
    .movePointerTo([0, 0.25])
    .halfEllipse(2, 0.5, 0.2)
    .close();

  gc();

  return shape.fuse(baseThreadSketch.loftWith(topThreadSketch));
};
```

**Reply by promontis** [reply]
Ah ok, so no opensource?

Your example looks really great btw! What does `r` and `gc` do in the `const [r, gc] = localGC();`? Also what happens with the `bottle` object afterwards? Is `shape` also a ThreeJS `Geometry`?

**Reply by sgenoud** [reply]
> Ah ok, so no opensource?

It depends on what you define as opensource! I plan to publish the code on github, I am currently thinking on the way to license it so that if people use it for a commercial purpose I can benefit from it as well.

> What does `r` and `gc` do in the `const [r, gc] = localGC();`

it is a way to flag objects for "garbage collection". You need to explicitely delete opencascade object, and my library tries to be smart and clean after itself. But sometimes there is no escape to deleting stuff explicitely. So this is a little helper to do just that.

> Also what happens with the bottle object afterwards? Is shape also a ThreeJS Geometry?

This is just a wrapper around an opencascade Shape object. It has some methods to make it easier to build a ThreeJS geometry out of it (or export it as STEP / STL). For now my aim is really to build a nice library to "draw" - what one does with it is another step.

**Reply by promontis** [reply]
Ah cool! Yeah, I really like your Sketcher / FaceSketcher API! You also mentioned cascadeQuery... tried to look it up, but couldn't find it. Is that also a library you built?

**Reply by sgenoud** [reply]
Sorry I meant CadQuery!

**Reply by promontis** [reply]
@sgenoud any progress on this? Would love to try it out!

**Reply by promontis** [reply]
Also, did you base it of the new Sketcher CadQuery has in PR? This one... https://github.com/CadQuery/cadquery/blob/sketch/cadquery/sketch.py

Seems to support parametric 2d sketching :)

**Reply by sgenoud** [reply]
> @sgenoud any progress on this? Would love to try it out!

Still working on it - currently in the progress of cleaning up the interfaces, typing and documenting them. 

My current plan is to publish it with AGPL and offer the possibility to buy exception to the rule.

> Seems to support parametric 2d sketching

While the parametric constraints are great in visual editing, I am not sure that they translate well in the case of code cad. This does not seem to be documented in cad query yet, is it?

**Reply by Irev-Dev** [reply]
I agree with constraints being hard to read and reason about when trying to code them.

**Reply by promontis** [reply]
@sgenoud Sounds great! Do you have an estimate on when you can publish the repo (hopefully soon)? Also, could you add a sponsor button? I'm sponsoring some other open source projects as well that I find cool and this is definitely one! Not too much btw (like $10 p/m), I'm no big company or anything like that. 

> This does not seem to be documented in cad query yet, is it?

The PR has already some docs in it.

**Reply by sgenoud** [reply]
I realise that I have forgotten to mention here that my library based on opencascadejs, `replicad` is published: https://replicad.xyz/

**Reply by zalo** [reply]
Very nice; I like your Sketching and Finder abstractions!

**Reply by donalffons** [reply]
This is fantastic! Congratulations on that library and the excellent documentation.
Think about requesting your project to be listed [here](https://dev.opencascade.org/about/projects_and_products) :slightly_smiling_face:.

Are you considering to release this as an NPM library, too?!

**Reply by sgenoud** [reply]
It is already on [npm](https://www.npmjs.com/package/replicad)! (the opencascade [custom build](https://www.npmjs.com/package/replicad-opencascadejs) as well, but it should work with the standard one as well)
